### DISCUSSION TITLE
Move all PositionZ for precise z-calibration?

### DISCUSSION THREAD

**Discussion by msspad** [sn4k3/UVtools#423] [Q&A] [answered]
1) A very precise z-calibration is essential when putting models directly on the build plate. Some printers offer to adjust the z-calibration with the touchscreen. It's generally putting the build plate at some 0.01 mm above the learned z-calibration. This option is missing in Phrozen Mini 4k/8k, is there an option in UVTools to move all PositionZ upwards by a constant?

2) I've been experimenting with fuzzy skin/corrode and noticed that pixels behind 255-intensity pixels get dimmed. Any reason for this except that it's easier to implement? (I don't believe subsurface scattering is a thing for resin models :-) The goal of corrode in UVTools is to achieve a matt finish of the surface, right?

3) typo in docs: "yeild"->yield

**Comment by sn4k3** [comment, marked as answer]
1. Never saw that function in a printer, setting a offset over calibrated value, often people would say, put another sheet and calibrate... 
But with UVtools is totally possible, via terminal or scripting, just iterate all layers and increase PositionZ by your offset value. This can be a tool to add in latter.
- Go to File - Terminal (Ctrl+Shift+T)
- Write:
```C#
const float offset = 0.01f;
foreach (var layer in SlicerFile) layer.PositionZ += offset;
```
- Send and close, they are now offset
2. All pixels, including 255 are dimmed if the lucky random gets a negative value, the corrode does not look at brightness levels.
Logic: https://github.com/sn4k3/UVtools/blob/4f3dd8b3a7597ce5a9b7b45d2d825bd4ab03e07f/UVtools.Core/Operations/OperationPixelArithmetic.cs#L795
If you want only to process the pure whites, set pixel threshold to **253**
The goal is a bit universal, corrode can do matt, fuzzy skin, bad texture or whatever you are aiming for, just need the correct parameters for each case. 
3. Where exactly?

**Reply by msspad** [reply]
1) Saw z offset here https://www.youtube.com/watch?v=wddKqQX2AOU but I'm also not familiar with it. 
It turns out that when using long WaitBeforeCure in the e.g. the first 20 layers, the height of the model becomes very repeatable. The accuracy with sheets+screws is totally fine for rafts but don't allow the highest accuracy possible now with WaitBeforeCure .
Great, will just assign an updated PositionZ in all layers with a script.

2) ok, thanks, I will try a few things with scripts.

3) OperationLayerReHeight.cs line 57 (sorry, haven't git around)

**Reply by sn4k3** [reply]
1. I think i can add that function to layer reheight as a new option
2. 👍
3. It's fixed, thanks

**Reply by sn4k3** [reply]
Here the draft, tell me if i'm missing some other option

![image](https://user-images.githubusercontent.com/113281/155432153-e45d9f2b-8e5b-4dcd-82d8-9a1ab3e802b8.png)

**Reply by msspad** [reply]
great work, thanks! I'm sure you also add a warning when the first PositionZ gets negative.

**Reply by sn4k3** [reply]
> great work, thanks! I'm sure you also add a warning when the first PositionZ gets negative.

Yup, it have some safe-checks
